_posts/2024_07_07-EOF.md

```diff
@@ -0,0 +1,101 @@
+
+I would really like to see the speed improvements of a mainnet block with and without these checks.
+
+## Address Space Extension
+
+EOF does not allow for EXTCALL, EXTDELEGATECALL or EXTSTATICCALL to call addresses where the upper 12 bytes are set. This restriction allows to set some of these in the future as a bitmask for address space extension.
+
+However this is not required in legacy contracts, which makes it impossible to move to address space extension for the legacy evm anyway.
+
+## Gas observability
+
+EOF also gets rid of the GAS opcode and replaces CALL, DELEGATECALL and STATICCALL with variants that do not allow the user to specify how much gas they would like to send with the call. This allows us to easier change gas prices for opcodes in the future.
+
+I'm pretty sure that getting rid of gas observability will impact a few users, they can still work around this by calling into a legacy contract which does the gas observation. 
+
+# Disadvantages
+
+Now that I've talked about the advantages of EOF (and admittedly already gave some of my thoughts about it), I want to discuss some of the disadvantages of EOF that I see.
+
+## Complexity
+
+The biggest issue I see is that EOF is extremely complex. It introduces RJUMP, RJUMPI, RJUMPV, CALLF, RETF, JUMPF, EOFCREATE, RETURNCONTRACT, DATALOAD, DATALOADN, DATASIZE, DATACOPY, DUPN, SWAPN, EXCHANGE, RETURNDATALOAD, EXTCALL, EXTDELEGATECALL and EXTSTATICCALL.
+
+All-in-all EOF introduces 19 new opcodes. Additionally it removes 16 opcodes from EOF and changes the behavior of multiple opcodes in the existing EVM. These changes have already resulted in two consensus issues in Besu and evmone (used by erigon) even before EOF made it to mainnet. Luckily these issues were found by Martin Holst Swende before they made it into a release.
+
+All these new opcodes must be tested, especially because some of these (like EOFCREATE and RETURNCONTRACT) interact with each other. 
+
+Additionally EOF introduces the code and stack validation. This means that the code is verified before it's being deployed. Therefore the code validity now becomes part of consensus which means that bugs in the code or stack validation will result in consensus bugs. Even worse, if the [code and stack validation spec](https://github.com/ipsilon/eof/blob/main/spec/eof.md#stack-validation) has a bug, a faulty contract could be deployed which would result in undefined behavior if called.
+
+## Impact analysis
+
+As far as I know, no impact analysis was done yet. While compiler teams have indicated that they like the changes, they have not published any reliable numbers on gas reductions, code size decrease or potential speed increases. 
+
+A change like EOF is really only useful if users (big protocols) will move away from their existing contracts and migrate over to EOF contracts. There need to be very good reasons for protocols to switch to a new contract though like significant gas cost reductions.
```

Comment:
I feel like they will move if there is any improvement in gas, so I'm not so worried on this.